### ISSUE TITLE
DEPRECATION WARNING

### ISSUE THREAD

**Issue by mibamur** [ActsAsParanoid/acts_as_paranoid#53]
```
DEPRECATION WARNING: alias_method_chain is deprecated. Please, use Module#prepend instead. From module, you can access the original method using super. (called from singleton class at /Users/user/.rvm/gems/ruby-2.3.1@rails5/bundler/gems/acts_as_paranoid-c2db19554dda/lib/acts_as_paranoid/associations.rb:6)
DEPRECATION WARNING: alias_method_chain is deprecated. Please, use Module#prepend instead. From module, you can access the original method using super. (called from block in included at /Users/user/.rvm/gems/ruby-2.3.1@rails5/bundler/gems/acts_as_paranoid-c2db19554dda/lib/acts_as_paranoid/preloader_association.rb:11)
```

**Comment by zzak** [comment]
@mibamur Could you submit a patch?

**Comment by mibamur** [comment]
@zzak I have a few experience in it
not now, later

**Comment by trcarden** [comment]
@zzak see: https://github.com/ActsAsParanoid/acts_as_paranoid/pull/55

**Comment by mibamur** [comment]
https://github.com/ActsAsParanoid/acts_as_paranoid/pull/54

**Comment by zzak** [comment]
@trcarden This patch is great if we want to drop backwards compatibility with older Ruby versions -- but at the moment I'd like to keep it for now.

Maybe we could submit that patch and start thinking about 1.0.0 w/ only modern versions of Ruby/Rails supported?

**Comment by zzak** [comment]
@mibamur LGTM, if it kills the warning and doesn't change backwards compatibility.

**Comment by mibamur** [comment]
@zzak 

could we change https://github.com/ActsAsParanoid/acts_as_paranoid/pull/54 like this for backwards compatibility?

```rb
        if Rails::VERSION::STRING.to_f < 5 || RUBY_VERSION.to_f < 2
          alias_method_chain :belongs_to, :deleted
        else
          alias_method :belongs_to_without_deleted, :belongs_to
          alias_method :belongs_to, :belongs_to_with_deleted
        end

```

**Comment by trcarden** [comment]
@zzak: Do you want me to move the gem's version to 1.0.0 in my patch set then and kill the old dependencies?

I think it makes sense to kill some of the older versions since Ruby 1.9 and 2.0 are now end of life 
* https://www.ruby-lang.org/en/news/2016/02/24/support-plan-of-ruby-2-0-0-and-2-1/
* https://www.ruby-lang.org/en/news/2014/01/10/ruby-1-9-3-will-end-on-2015/

Also going along with #54 if you take a look at what alias_method_chain did under the covers ([see here](https://github.com/rails/rails/blob/628474f2d3330f8a5f14d87153efbed1fcb23e0e/activesupport/lib/active_support/core_ext/module/aliasing.rb#L26-L48)) that is just copy-pasting in the same deprecated functionality. If you are going to do that you might as well just copy and paste `alias_method_chain` in a lib directory and remove the deprecation notice so that you get all the benefits of the original method call (they differ slightly from just two alias_method calls)

Finally here are some reasons to use the new prepend:
http://www.justinweiss.com/articles/rails-5-module-number-prepend-and-the-end-of-alias-method-chain/

**Comment by zzak** [comment]
It's trivial to maintain compatibility (one case statement), so I'd rather not drop support or bump major version at this time.

**Comment by trcarden** [comment]
That's too bad. I don't think supporting ruby versions that are dead is worth it. Also this is a mature gem so bumping a major version isn't going to leave anyone out of new features. That said I am not going to maintain it so maybe it will make your life easier. I will keep my fork up there if people wanted to eventually make this gem use more modern techniques

**Comment by mibamur** [comment]
could someone answer
is `if Rails::VERSION::STRING.to_f < 5 || RUBY_VERSION.to_f < 2` one of possible way?

**Comment by BrianHawley** [comment]
https://bugs.ruby-lang.org/issues/10847 is a problem with stack overflows with `prepend`. It was fixed in Ruby 2.1.6 and 2.2.2, but not in any prior Ruby versions. You should check the earliest version of Ruby supported for use with the earliest version of Rails that you support, and write the code that works with that. Unless you can be sure that your code isn't affected, of course.

**Comment by MrWilhelm** [comment]
Anything new about updating ? It's annoying and been months. Really.

**Comment by zzak** [comment]
For the record, this is just a `WARNING` there is no need to take immediate action unless there is an actual bug.

As for getting this fixed, I already commented how I'd like to resolve this, anyone is free to submit a patch!

**Comment by nneufeld** [comment]
@zzak Is there any reason https://github.com/ActsAsParanoid/acts_as_paranoid/pull/54 can't be merged? Because it doesn't use `prepend`, it doesn't break compatibility, but it does get rid of the deprecation warning.

**Comment by estebanbouza** [comment]
@zzak Could you comment on @nneufeld 's statement before? Is the gem abandoned?

**Comment by zzak** [comment]
@estebanbouza @nneufeld Totally, sorry I just missed this comment.

Thanks for your patience!

**Comment by noise2** [comment]
@zzak Why this issue still open?? @nneufeld's solution will fix the problem, why not merge it?!

**Comment by zzak** [comment]
Merged #54
